File: 33-search-in-rotated-sorted-array/33-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& arr, int target) {
        int l=0,r=arr.size()-1;
        while(l<=r){
            int mid = l + (r-l)/2;
            if(arr[mid]==target)
                return mid;
            else if(arr[l]<=arr[mid]){
                if(arr[l]<=target and target <arr[mid])
                    r=mid-1;
                else
                    l=mid+1;
            }   
            else{
                if(arr[mid]<target and target<=arr[r])
                    l=mid+1;
                else
                    r=mid-1;
            }
        }
        return -1;
    }
};

File: 33-search-in-rotated-sorted-array/33-search-in-rotated-sorted-array.py
class Solution:
    def search(self, A: List[int], target: int) -> int:
        left, right = 0, len(A) - 1
        while left <= right:
            mid = left + (right - left) // 2
            if A[mid] == target: return mid

            # inflection point is to the right of mid i.e. Left part is strictly increasing
            if A[mid] >= A[left]:
                if A[left] <= target < A[mid]:
                    right = mid - 1
                else:
                    left = mid + 1

            # inflection point is to the left of mid i.e. Right part is strictly increasing
            else:
                if A[mid] < target <= A[right]:
                    left = mid + 1
                else:
                    right = mid - 1

        return -1